torchaudio/prototype/ctc_decoder/ctc_decoder.py

```diff
@@ -0,0 +1,224 @@
+            _, score = self.lm.score(start_state, word_idx)
+            for spelling in spellings:
+                spelling_idx = [self.tokens_dict.get_index(token) for token in spelling]
+                self.trie.insert(spelling_idx, word_idx, score)
+        self.trie.smear(_SmearingMode.MAX)
+
+        self.decoder = _LexiconDecoder(
+            self.decoder_options,
+            self.trie,
+            self.lm,
+            self.silence,
+            self.blank,
+            self.unk_word,
+            [],
+            False,  # word level LM
+        )
+
+    def _get_tokens(self, idxs: torch.IntTensor) -> torch.LongTensor:
+        idxs = (g[0] for g in it.groupby(idxs))
+        idxs = filter(lambda x: x != self.blank, idxs)
+        return torch.LongTensor(list(idxs))
+
+    def decode(
+        self,
+        emissions: torch.FloatTensor,
+        lengths: Optional[torch.Tensor] = None
+    ) -> List[List[Dict]]:
+        """
+        Args:
+            emissions (FloatTensor): tensor of shape `(batch, frame, num_tokens)` storing sequences of
+                probability distribution over labels; output of acoustic model
+            lengths (Tensor or None, optional): tensor of shape `(batch, )` storing the valid length of
+                in time axis of the output Tensor in each batch
+
+        Returns:
+            List[Hypothesis]:
+                List of sorted best hypotheses for each audio sequence in the batch.
+
+                Each hypothesis is named tuple with the following fields:
+                    tokens: torch.LongTensor of raw token IDs
+                    score: hypothesis score
+                    words: list of decoded words
+        """
+        B, T, N = emissions.size()
+        if lengths is None:
+            lengths = torch.full((B, ), T)
+
+        hypos = []
+        for b in range(B):
+            emissions_ptr = emissions.data_ptr() + 4 * b * emissions.stride(0)
+
+            results = self.decoder.decode(emissions_ptr, lengths[b], N)
+
+            nbest_results = results[: self.nbest]
+            hypos.append(
+                [
+                    Hypothesis(
+                        self._get_tokens(result.tokens),  # token ids
+                        list(self.word_dict.get_entry(x) for x in result.words if x >= 0),  # words
+                        result.score,  # score
+                    )
+                    for result in nbest_results
+                ]
+            )
+
+        return hypos
+
+    def idxs_to_tokens(self, idxs: torch.LongTensor) -> List:
+        """
+        Map raw token IDs into correponding tokens
+
+        Args:
+            idxs (LongTensor): raw token IDs generated from decoder
+
+        Returns:
+            List: tokens corresponding to the input IDs
+        """
+        return [self.tokens_dict.get_entry(idx.item()) for idx in idxs]
+
+
+def kenlm_lexicon_decoder(
+    lexicon_file: str,
+    tokens_file: str,
+    kenlm_file: str,
```

Comment:
I think tokens can be more flexible than requiring a file. Tokens should be able to be directly fed as list of str, so that it can accept the value from `w2v2bundle.get_labels()`.

Similarly I feel that lexicon argument can be a regular dict as well. Looking at the implementation of `_Dictionary`, it needs reverse look up which can be constructed from regular dict if I'm not mistaken. Not requiring a file gives better flexibility at testing but I'm not entirely sure if that's best for UX. Let's follow-up on this one as we add pretrained model support and write tutorials.

Reply:
I will follow up on both comments in a future PR, since the way this is currently being done would require C++ changes to the `_Dictionary` binding or perhaps transitioning to use a different Python/PyTorch based Dictionary